backend/src/impact.rs

```diff
@@ -37,19 +33,48 @@ impl Impact {
 
     /// Returns a feature per road, with `before` and `after` counts, and a `max_count` foreign
     /// member
-    pub fn recalculate(&mut self, map: &MapModel) -> FeatureCollection {
-        if self.counts_before.is_empty() {
-            info!("Calculating impacts before edits");
-            self.counts_before = map.router_before.od_to_counts(&self.requests);
+    pub fn recalculate(&mut self, map: &MapModel, fast_sample: bool) -> FeatureCollection {
+        // Which requests are we using?
+        let requests = if fast_sample {
+            if self.sampled_requests.is_empty() {
+                info!("Calculating a fast sample of requests");
+                self.sampled_requests = match &map.demand {
+                    Some(demand) => demand.make_requests(map, fast_sample),
+                    None => od::synthetic_od_requests(map),
+                };
+            }
+            &self.sampled_requests
+        } else {
+            if self.all_requests.is_empty() {
+                info!("Calculating all requests");
+                self.all_requests = match &map.demand {
+                    Some(demand) => demand.make_requests(map, fast_sample),
+                    None => od::synthetic_od_requests(map),
+                };
+            }
+            &self.all_requests
+        };
+
+        if self.counts_before.is_empty() || fast_sample != self.last_fast_sample {
+            info!(
+                "Calculating impacts before edits ({} requests)",
+                requests.len()
+            );
+            self.counts_before = map.router_before.od_to_counts(requests);
+            self.last_fast_sample = fast_sample;
```

Comment:
Will this assignment throw off the check 2 lines down? 

My guess is that self.last_fast_sample should be set at the very end of this method.

Reply:
AHHH. Good catch, thank you!